crates/topos-p2p/src/behaviour/grpc.rs

```diff
@@ -0,0 +1,516 @@
+        // If there is a pending outbound connection for this peer
+        // return the request id
+        if let Some(request) = self.pending_outbound_connections.get(peer_id) {
+            return OutboundConnection::Pending {
+                request_id: request.request_id,
+            };
+        }
+
+        if let Some(connections) = self.connected.get_mut(peer_id) {
+            match connections.first() {
+                Some(Connection {
+                    id,
+                    address,
+                    request_id: Some(request_id),
+                    channel: Some(channel),
+                }) => OutboundConnection::Connected(OutboundConnectedConnection {
+                    request_id: *request_id,
+                    channel: channel.clone(),
+                }),
+                Some(Connection {
+                    id,
+                    address,
+                    request_id: Some(request_id),
+                    channel,
+                }) => {
+                    debug!("Peer already connected but no channel bound");
+
+                    OutboundConnection::Pending {
+                        request_id: *request_id,
+                    }
+                }
+                Some(_) => self.open_connection(peer_id),
+                _ => {
+                    debug!("No connection for this peer {}", peer_id);
+                    self.open_connection(peer_id)
+                }
+            }
+        } else {
+            debug!("Buffering sender as no available connection to peer {peer_id} yet");
+            self.open_connection(peer_id)
+        }
+    }
+
+    /// Return the next outbound request id
+    fn next_request_id(&mut self) -> RequestId {
+        let request_id = self.next_request_id;
+        self.next_request_id.0 += 1;
+
+        request_id
+    }
+
+    /// Try to open a connection with the given peer.
+    fn open_connection(&mut self, peer_id: &PeerId) -> OutboundConnection {
+        info!("Opening gRPC outbound connection to peer {peer_id}");
+
+        let (notifier, receiver) = oneshot::channel();
+        let request_id = self.next_request_id();
+
+        self.pending_outbound_connections
+            .entry(*peer_id)
+            .or_insert_with(|| OutboundConnectionRequest {
+                request_id,
+                notifier,
+            });
+
+        self.pending_events.push_back(ToSwarm::Dial {
+            opts: DialOpts::peer_id(*peer_id).build(),
+        });
+
+        OutboundConnection::Opening {
+            request_id,
+            receiver,
+        }
+    }
+
+    /// Handle the [`ConnectionEstablished`] event comming from the [`Swarm`]
+    /// and try to open a gRPC channel using a [`ConnectionHandler`].
+    fn on_connection_established(
+        &mut self,
+        ConnectionEstablished {
+            peer_id,
+            connection_id,
+            endpoint,
+            failed_addresses,
+            other_established,
```

Comment:
I'm not sure I understand what this value means. Other connections with this same peer?

Reply:
Yes, it's the number of existing established connection with this peer